### PR TITLE
[WIP] DS Proof implementation

### PR DESCRIPTION
**Pull request by ptschip**
Taking Tom Zanders DS Proof PR and working on top of that to get it to work with BU.  Also want to hook it into the already existing double spend framework from @dagurval.  We also need a proper set of tests for this.

### CODE REVIEW

**Comment by ptschip** [comment]
@Justaphf not sure if you still around but this PR will be the one that gets merged for DS proofs.  It's basically ready for testing if you want to give it a try.  I just need to build some python tests and get this PR reviewed by the team.

**Comment by Justaphf** [comment]
@ptschip I'll try to create a build over the weekend and play around with it.  I've been running the code off PR 2073, periodically rebased on dev (latest rebase is about 10 commits behind current dev branch).  Though mostly I've just been running it and occasionally looking at the logs for DSP messages, not really testing it beyond noting if it's crashing.  I believe the original crash I had reported was fixed by e737372.

In 2abb165 you change the message name, so won't this break compatibility with the Flowee nodes on the network.  I've been ensuring my node connects to a couple of Flowee nodes so the receipt of DSP notifications gets exercised, not just the local creation of them.  Since they are still using the beta tag should we support both, at least temporarily, for processing received messages?

**Comment by ptschip** [comment]
@Justaphf not really sure what to do with that at this point...seems kind of silly to have a dsproof-beta tag within a message especially since we use XVERSION for things like this but for now I'll just leave it in this PR as dsproof.  I'd hate to have to support both message types both in and out but we could do it if we must.  For now, if you need to get messages from tomz's nodes then perhaps just change it back to dsproof-beta locally.

**Comment by Justaphf** [comment]
I think what I'll do is just modify so my test node will still understand receipt of the `-beta` message to mean the same thing, but not send it with the `-beta` suffix.  Should just be a 1 line change.  Based on the wording of the spec document it looks like Tom's intention is to remove the `-beta` suffix from the message once the specification is finalized anyway.

File: src/DoubleSpendProof.cpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    if (scriptType == P2PKH)
+    {
+        inScript << m_spender1.pushData.front();
+        inScript << pubkey;
+    }
+    DSPSignatureChecker checker1(this, m_spender1, amount);
+    ScriptError_t error;
+    if (!VerifyScript(inScript, prevOutScript, 0 /*flags*/, MAX_OPS_PER_SCRIPT, checker1, &error))
+    {
+        LOGA("DoubleSpendProof failed validating first tx due to %s\n", ScriptErrorString(error));
```
Comment (Justaphf) [inline src/DoubleSpendProof.cpp]: Should use `LOG(DSPROOF, ...)` instead of `LOGA()`

File: src/DoubleSpendProof.cpp [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    inScript.clear();
+    if (scriptType == P2PKH)
+    {
+        inScript << m_spender2.pushData.front();
+        inScript << pubkey;
+    }
+    DSPSignatureChecker checker2(this, m_spender2, amount);
+    if (!VerifyScript(inScript, prevOutScript, 0 /*flags*/, MAX_OPS_PER_SCRIPT, checker2, &error))
+    {
+        LOGA("DoubleSpendProof failed validating second tx due to %s\n", ScriptErrorString(error));
```
Comment (Justaphf) [inline src/DoubleSpendProof.cpp]: Should use `LOG(DSPROOF, ...)` instead of `LOGA()`

File: src/DoubleSpendProof.cpp (hunk 1 above)
Comment (ptschip) [reply src/DoubleSpendProof.cpp]: done

File: src/DoubleSpendProof.cpp (hunk 2 above)
Comment (ptschip) [reply src/DoubleSpendProof.cpp]: done

**Comment by sickpig** [comment]
@ptschip @Justaphf since we moved to gitLlab for dev it would be great to continue dev there

**Pull request closed without merging**